packages/core/src/util.ts

```diff
@@ -0,0 +1,939 @@
+
+export type InitInterfaceParams<T extends schemas.$ZodLooseShape, Extra extends Record<string, unknown>> = Identity<{
+  optional: OptionalInterfaceKeys<keyof T>;
+  defaulted: DefaultedInterfaceKeys<keyof T>;
+  extra: Extra;
+}>;
+
+export type MergeOptional<A extends schemas.$ZodInterface, B extends schemas.$ZodInterface> =
+  | Exclude<A["_zod"]["optional"], keyof B["_zod"]["def"]["shape"]>
+  | B["_zod"]["optional"];
+export type MergeDefaulted<A extends schemas.$ZodInterface, B extends schemas.$ZodInterface> =
+  | Exclude<A["_zod"]["defaulted"], keyof B["_zod"]["def"]["shape"]>
+  | B["_zod"]["defaulted"];
+
+export type MergeInterfaceParams<
+  A extends schemas.$ZodInterface,
+  B extends schemas.$ZodInterface,
+  // BKeys extends PropertyKey,
+> = Identity<{
+  optional: Exclude<A["_zod"]["optional"], keyof B["_zod"]["def"]["shape"]> | B["_zod"]["optional"];
+  defaulted: Exclude<A["_zod"]["defaulted"], keyof B["_zod"]["def"]["shape"]> | B["_zod"]["defaulted"];
+  extra: A["_zod"]["extra"];
+}>;
+
+export type ExtendInterfaceParams<A extends schemas.$ZodInterface, Shape extends schemas.$ZodLooseShape> = Identity<{
+  optional: Exclude<A["_zod"]["optional"], CleanKeys<keyof Shape>> | OptionalInterfaceKeys<keyof Shape>;
+  defaulted: Exclude<A["_zod"]["defaulted"], CleanKeys<keyof Shape>> | DefaultedInterfaceKeys<keyof Shape>;
+  extra: A["_zod"]["extra"];
+}>;
+export type ExtendInterfaceShape<A extends schemas.$ZodLooseShape, B extends schemas.$ZodLooseShape> = Extend<
+  A,
+  CleanInterfaceShape<B>
+>;
+export type InterfaceParams<T extends schemas.$ZodInterface> = {
+  optional: T["_zod"]["optional"];
+  defaulted: T["_zod"]["defaulted"];
+  extra: T["_zod"]["extra"];
+};
+
+export function cleanInterfaceKey(key: string): string {
+  return key.replace(/^\?/, "").replace(/\?$/, "");
+}
+
+export function cleanInterfaceShape<T extends schemas.$ZodLooseShape>(
+  _shape: T
+): {
+  shape: CleanInterfaceShape<T>;
+  keyMap: Record<string, string>;
+  optional: string[];
+  defaulted: string[];
+} {
+  const keyMap: Record<string, string> = {};
+  const shape = {} as CleanInterfaceShape<T>;
+  const optional: string[] = [];
+  const defaulted: string[] = [];
+
+  for (const [key, value] of Object.entries(_shape)) {
+    if (key.endsWith("?")) optional.push(key.slice(0, -1));
+    if (key.startsWith("?")) defaulted.push(key.slice(1));
+    const cleanKey = cleanInterfaceKey(key);
+    (shape as any)[cleanKey] = value;
+    keyMap[cleanKey] = key;
+  }
+
+  return { shape, keyMap, optional, defaulted };
+}
+
+export const NUMBER_FORMAT_RANGES: Record<checks.$ZodNumberFormats, [number, number]> = {
+  safeint: [Number.MIN_SAFE_INTEGER, Number.MAX_SAFE_INTEGER],
+  int32: [-2147483648, 2147483647],
+  uint32: [0, 4294967295],
+  float32: [-3.4028234663852886e38, 3.4028234663852886e38],
+  float64: [-1.7976931348623157e308, 1.7976931348623157e308],
```

Comment:
there is a const for that
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/MAX_VALUE

```suggestion
  float64: [-Number.MAX_VALUE, Number.MAX_VALUE],
```

Reply:
Would you mind putting in a PR with these changes? I think they're all good.

Reply:
https://github.com/colinhacks/zod/pull/4222